File: all_banchmark/05_ackermann/ackermann.c
// Ackermann Function Benchmark - C
// Tests deep recursion performance (reduced input to avoid stack overflow)

#include <stdio.h>
#include <stdint.h>

int64_t ackermann(int64_t m, int64_t n) {
    if (m == 0) return n + 1;
    if (n == 0) return ackermann(m - 1, 1);
    return ackermann(m - 1, ackermann(m, n - 1));
}

int main() {
    // ackermann(3, 10) = 8189 (manageable recursion depth)
    int64_t m = 3;
    int64_t n = 10;
    int64_t result = ackermann(m, n);
    printf("ackermann(%lld, %lld) = %lld\n", m, n, result);
    return 0;
}

File: all_banchmark/05_ackermann/ackermann.rs
// Ackermann Function Benchmark - Rust
// Tests deep recursion performance (reduced input to avoid stack overflow)

fn ackermann(m: i64, n: i64) -> i64 {
    if m == 0 {
        n + 1
    } else if n == 0 {
        ackermann(m - 1, 1)
    } else {
        ackermann(m - 1, ackermann(m, n - 1))
    }
}

fn main() {
    // ackermann(3, 10) = 8189 (manageable recursion depth)
    let m: i64 = 3;
    let n: i64 = 10;
    let result = ackermann(m, n);
    println!("ackermann({}, {}) = {}", m, n, result);
}
